### ISSUE TITLE
Mysterious overlap error

### ISSUE THREAD

**Issue by masaeedu** [purescript/purescript#3814]
I'm trying to define the following:

```purescript
module Tensor where

import Prelude

import Data.Bifunctor (class Bifunctor)
import Data.Either (Either(..), either)
import Data.Tuple (Tuple(..), fst, snd)
import Data.Tuple.Nested ((/\))

type Iso p a b = { fwd :: p a b, bwd :: p b a }

class Bifunctor t <= Associative t p
  where
  assoc :: forall a b c. Iso p (t a (t b c)) (t (t a b) c)

instance associativeCartesian :: Associative Tuple (->)
  where
  assoc =
    { fwd: \(a /\ b /\ c) -> (a /\ b) /\ c
    , bwd: \((a /\ b) /\ c) -> a /\ b /\ c
    }

instance associativeCocartesian :: Associative Either (->)
  where
  assoc =
    { fwd: either (Left <<< Left) (either (Left <<< Right) Right)
    , bwd: either (either Left (Right <<< Left)) (Right <<< Right)
    }

class Associative t p <= Tensor t i p | p i -> t, p t -> i
  where
  lunit :: forall a. Iso p (t i a) a
  runit :: forall a. Iso p (t a i) a

instance tensorCartesian :: Tensor Tuple Unit (->)
  where
  lunit = { fwd: snd, bwd: Tuple unit }
  runit = { fwd: fst, bwd: flip Tuple unit }

instance tensorCocartesian :: Tensor Either Void (->)
  where
  lunit = { fwd: either absurd identity, bwd: Right }
  runit = { fwd: either identity absurd, bwd: Left }
```

However I get an error on the last line:

```
[PureScript OverlappingInstances] [E]   Overlapping type class instances found for

    Data.Category.Tensor.Tensor Either
                                Void
                                Function

  The following instances were found:

    Data.Category.Tensor.tensorCartesian
    Data.Category.Tensor.tensorCocartesian


in type class instance

  Data.Category.Tensor.Tensor Either
                              Void
                              Function
```

Given that there's neither a pair of `p i` for which there are multiple instances of `Tensor p i _`, nor a pair of `p t` for which there are multiple instances of `Tensor p _ t`, I don't understand what the problem is.

**Comment by hdgarrood** [comment]
Thanks for reporting. This does seem like a bug to me.

**Comment by MonoidMusician** [comment]
Fixed by #4195 😄
